### PR TITLE
[#30083][prism] Stabilize additional teststream cases.

### PR DESCRIPTION
**Pull request by lostluck**
While working on adding Processing Time handling, it came to light that there were some inconsistent cases WRT handling TestStream based pipelines, in particular when there were no impulses included.

This fix uses the existing watermark hold system to restrain execution until the test stream is ready, and reflect holds based on test stream watermark events properly before their execution.

This discovered flaky fail behavior due to how timers and data outputs were handled in bundle processing, also fixed in this PR. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/go/test/integration/primitives/teststream_test.go [hunk 1]
```diff
@@ -71,3 +71,23 @@ func TestTestStreamTwoUserTypeSequences(t *testing.T) {
 	integration.CheckFilters(t)
 	ptest.BuildAndRun(t, TestStreamTwoUserTypeSequences)
 }
+
+func TestTestStreamSimple(t *testing.T) {
+	integration.CheckFilters(t)
+	ptest.BuildAndRun(t, TestStreamSimple)
+}
+
+func TestTestStreamTestStreamSimple_InfinityDefault(t *testing.T) {
```
Comment (riteshghorse) [inline sdks/go/test/integration/primitives/teststream_test.go:80]:
```suggestion
func TestTestStreamSimple_InfinityDefault(t *testing.T) {
```

File: sdks/go/test/integration/primitives/teststream_test.go [hunk 2]
```diff
@@ -71,3 +71,23 @@ func TestTestStreamTwoUserTypeSequences(t *testing.T) {
...
+	integration.CheckFilters(t)
+	ptest.BuildAndRun(t, TestStreamSimple_InfinityDefault)
+}
+
+func TestTestStreamToGBK(t *testing.T) {
+	integration.CheckFilters(t)
+	ptest.BuildAndRun(t, TestStreamToGBK)
+}
+
+func TestTestStreamTimersEventTimeTestStream(t *testing.T) {
```
Comment (riteshghorse) [inline sdks/go/test/integration/primitives/teststream_test.go:90]:
```suggestion
func TestTestStreamTimersEventTime(t *testing.T) {
```

File: sdks/go/test/integration/primitives/teststream.go [hunk 3]
```diff
@@ -172,3 +175,74 @@ func TestStreamInt16Sequence(s beam.Scope) {
...
 }
+
+// panicIfNot42 panics if the value is not 42.
+func panicIfNot42(v int) {
+	if v != 42 {
+		panic(fmt.Sprintf("got %v, want 42", v))
+	}
+}
+
+// dropKeyEmitValues panics if the value is not 42.
```
Comment (riteshghorse) [inline sdks/go/test/integration/primitives/teststream.go:186]:
```suggestion
// dropKeyEmitValues drops the key and emits the value.
```

**Review by riteshghorse — APPROVED** [review]
LGTM with a few nits, thanks!

**Comment by lostluck** [comment]
Nice catches! Thanks. Merging once the recommits pass again.

**Pull request merged**